### PR TITLE
[4.0] Separate sources from the distributable assets

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue # .

### Summary of Changes
This PR moves the static assets sources to the build directory.
There are 2 folders:
- media is holding all the legacy (eg ES5 js and css that don't have a relevant scss file)
- media_src is holding all the ES6 scripts and SCSS files
- the source for the templates remains in their respected template folders

There is no media folder anymore in the source and thus any change in this directory will not be flagged in git.

There is one more step now in the build process: creating the SPA for media manager

### Testing Instructions
Before fetching this branch make a copy of the folder media somewhere.
after fetch/pull you should run npm i and compare the 2 directories.
(the template assets weren't moved so nothing to check there!)

### Expected result
All assets regenerated with `npm i`, stop tracking compiled/minified files

### Documentation Changes Required
All the sources for the static assets are held in the `/build` folder.

PS. the `--compile-js` argument **will not** minify css and js files in the legacy folder for this you'll need to use the `--update` argument

### CODE REVIEW

**Comment by mbabker** [comment]
> There is one more step now in the build process: creating the SPA for media manager

Just make sure that step is included in `build/build.php` if it's not something coming from the default `npm install` that we're already doing.

**Comment by dgrammatiko** [comment]
@mbabker it's baked in the `npm i` so there shouldn't be a need to do something in the build.php

**Comment by mbabker** [comment]
I saw that after the comment, and actually there would be one change needed.  Adding a cleanup step to remove `administrator/components/com_media/node_modules` (after [this line](https://github.com/joomla/joomla-cms/blob/78f76411d2720da188d163c912bb24dab54f81fe/build/build.php#L211)) so it doesn't get packaged.

**Comment by dgrammatiko** [comment]
@mbabker done!

File: build.js [hunk 1]
```diff
@@ -22,11 +22,11 @@ const Program = require('commander');
...
 const installer = require('./build/build-modules-js/installation');
-const update = require('./build/build-modules-js/update');
-const css = require('./build/build-modules-js/compilescss');
-const Js = require('./build/build-modules-js/compilejs');
-const CEjs = require('./build/build-modules-js/compilecejs');
-const fixVend = require('./build/build-modules-js/minify-vendor');
+const copyAssets = require('./build/build-modules-js/update');
+const compileCSS = require('./build/build-modules-js/compilescss');
+const compileJS = require('./build/build-modules-js/compilejs');
+const WebComponents = require('./build/build-modules-js/compilecejs');
```
Comment (wilsonge) [inline build.js]: `compileWebcomponentsJs` please

File: build.js [hunk 2]
```diff
@@ -22,11 +22,11 @@ const Program = require('commander');
...
+const fixVendor = require('./build/build-modules-js/minify-vendor');
```
Comment (wilsonge) [inline build.js]: `minifyVendor` (or similar)

File: build/build-modules-js/compilescss.js [hunk 3]
```diff
@@ -77,6 +78,17 @@ const compileFiles = (options, path) => {
         cleaner.process(result.css.toString(), {from: undefined})
           .then(cleaned => prefixer.process(cleaned.css, {from: undefined}))
           .then((res) => {
+            // Ensure the folder exists or create it
+            const currentDir = Path.dirname(cssFile);
+            try{
+              fs.lstatSync(currentDir).isDirectory()
```
Comment (wilsonge) [inline build/build-modules-js/compilescss.js:84]: If it already exists we should probably ensure it is empty?

**Review by wilsonge — CHANGES_REQUESTED** [review]
Minor things but overall looks good

**Comment by wilsonge** [comment]
Final thing - we can't really drop the template CSS + minified CSS until we either .gitignore it or move it to another location like you proposed as it will always get regenerated when people compile the code...

File: build.js (hunk 1 above)
Comment (dgrammatiko) [reply build.js]: Probably `compileWebcomponents` the `Js` part is misleading as true web components also encapsulate the css

File: build/build-modules-js/compilescss.js (hunk 3 above)
Comment (dgrammatiko) [reply build/build-modules-js/compilescss.js:84]: Not necessarily, both the js transpiler and the css compiler will overwrite the given files. Also, the whole process assumes that someone begins from a  clean `git clone ...` and even in the case eg someone renames a file, the source will be committed and only locally on the developers' machine, there will be a redundant file in the media folder. Probably we need one more command: `cleanDist` or something similar that does some sane cleanup

**Pull request merged**